src/main/kotlin/org/openbase/bco/device/hass/manager/service/BlindStateServiceImpl.kt

```diff
@@ -31,12 +33,33 @@ import java.util.concurrent.Future
 
 class BlindStateServiceImpl<ST>(unit: ST) : HassService<ST>(unit),
     BlindStateOperationService where ST : BlindStateOperationService, ST : Unit<*> {
-    override fun setBlindState(blindState: BlindState): Future<ActionDescription> {
-        return setState(blindState)
-    }
+    override fun setBlindState(blindState: BlindState): Future<ActionDescription> =
+        callService(
+            hassServiceType =
+                when (blindState.value) {
+                    BlindState.State.UP -> HassServiceType.OPEN_COVER
+                    BlindState.State.DOWN -> HassServiceType.CLOSE_COVER
+                    BlindState.State.STOP -> HassServiceType.STOP_COVER
+                    else -> HassServiceType.UNKNOWN
+                },
+            state = blindState,
+        )
 
     @Throws(NotAvailableException::class)
-    override fun getBlindState(): BlindState {
-        return unit.blindState
+    override fun getBlindState(): BlindState = unit.blindState
+}
+
+fun HassStateDto.toBlindState(): BlindState.Builder {
+    return  BlindState.newBuilder().apply {
+        value = when (state) {
+            HassStateDto.STATE_OPENING,
+            HassStateDto.STATE_CLOSING,
+            HassStateDto.STATE_CLOSED -> BlindState.State.DOWN
+            HassStateDto.STATE_STOPPED -> BlindState.State.STOP
+            HassStateDto.STATE_OPEN -> BlindState.State.UP
+            else -> BlindState.State.UNKNOWN
+        }
+        openingRatio = 1 - (position?.let { it / 100.0 }
```

Comment:
```suggestion
        openingRatio = 1.0 - (position?.let { it / 100.0 }
```
this should improve the performance by avoiding the cast to double